File: app/src/main/java/org/jaaksi/pickerview/demo/BaseFragment.java
package org.jaaksi.pickerview.demo;

import android.app.Activity;
import android.app.Fragment;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * 创建时间：2018年02月12日13:56 <br>
 * 作者：fuchaoyang <br>
 * 描述：
 */

public abstract class BaseFragment extends Fragment {
  protected Activity mActivity;

  @Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
    Bundle savedInstanceState) {
    mActivity = getActivity();
    View view = inflater.inflate(getLayoutId(), container, false);
    view.setBackgroundColor(Color.WHITE);
    view.setClickable(true);
    initView(view);
    return view;
  }

  protected abstract int getLayoutId();

  protected abstract void initView(View view);

  protected void openFragment(Fragment fragment) {
    getFragmentManager().beginTransaction()
      .add(android.R.id.content, fragment)
      .addToBackStack(null)
      .commit();
  }
}

File: app/src/main/java/org/jaaksi/pickerview/demo/base/BaseFragment.kt
package org.jaaksi.pickerview.demo.base

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.viewbinding.ViewBinding
import java.lang.reflect.ParameterizedType

open class BaseFragment<T : ViewBinding> : Fragment() {
    lateinit var binding: T
    private var hasLoaded = false

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val type = javaClass.genericSuperclass as ParameterizedType
        val aClass = type.actualTypeArguments[0] as Class<*>
        val method = aClass.getDeclaredMethod(
            "inflate",
            LayoutInflater::class.java,
            ViewGroup::class.java,
            Boolean::class.java
        )
        binding = method.invoke(null, layoutInflater, container, false) as T
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initData(savedInstanceState)
        initView(savedInstanceState)
        binding.root.setBackgroundColor(Color.WHITE)
        binding.root.isClickable = true
        observer()
    }

    open fun initData(savedInstanceState: Bundle?) {}
    open fun initView(savedInstanceState: Bundle?) {}
    open fun observer() {}

    override fun onResume() {
        super.onResume()
//        if (!isHidden) {
            onVisibleChanged(true, !hasLoaded)
            hasLoaded = true
//        }

    }

    override fun onPause() {
        super.onPause()
        onVisibleChanged(visible = false, isLazy = false)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hasLoaded = false
    }

    /**
     * ViewPager和ViewPager2中都使用了setMaxLifecycle，onResume和onPause就是可见/不可见
     */
    open fun onVisibleChanged(visible: Boolean, isLazy: Boolean) {}

}